shared/ws/mirror.go

```diff
@@ -32,29 +31,20 @@ func MirrorRead(ctx context.Context, conn *websocket.Conn, rc io.ReadCloser) cha
 
 	go func() {
 		defer close(chDone)
-		_, _ = io.Copy(connRWC, rc)
 
-		// Send write barrier.
-		connRWC.Close()
+		_, _ = io.Copy(connRWC, rc)
 
 		logger.Debug("Websocket: Stopped read mirror", logger.Ctx{"address": conn.RemoteAddr().String()})
-	}()
 
-	go func() {
-		// Handle cancelation.
-		select {
-		case <-ctx.Done():
-			// Close the ReadCloser on cancel.
-			rc.Close()
-		case <-chDone:
-		}
+		// Send write barrier.
+		connRWC.Close()
```

Comment:
A small nit, but we are intentionally ignoring error here, so we could use `_ = connRWC.Close()`.

Otherwise, looks good to me.